### PR TITLE
[homematic] Solves stability issues with HmIP devices

### LINKED ISSUE #7762: [homematic ipwired] No function after a few days
@MHerbst  
after I have done the Update to 2.5.5 I have problems with Homematic IP Wired.

After rebooting openHab everything ist OK.
A few days later the HomematicWired things doesn't work.

Maybe there is a context to #7473 

99289294_10221902480380948_8498322394115473408_o

### PR DESCRIPTION
**Pull request by MHerbst**
The CCU gets unresponsive if several HmIP devices are installed and
always "non-blocking" requests are used. By reducing the use to what is
absolutely necessary, the problem can be avoided.

Fixes #7762

Signed-off-by: Martin Herbst <develop@mherbst.de>

### CODE REVIEW

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java [hunk 1]
```diff
@@ -68,57 +69,67 @@ protected String getRpcCallbackUrl() {
...
+        Exception reason = null;
+        for (int rpcRetryCounter = 1; rpcRetryCounter <= MAX_RPC_RETRY; rpcRetryCounter++) {
+            try {
+                byte[] response = send(port, request);
+                Object[] data = new XmlRpcResponse(new ByteArrayInputStream(response), config.getEncoding())
+                        .getResponseData();
+                return new RpcResponseParser(request).parse(data);
+            } catch (UnknownRpcFailureException | UnknownParameterSetException ex) {
+                throw ex;
+            } catch (Exception ex) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java:81]: Is it absolutely necessary to catch `Exception` here?

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java [hunk 2]
```diff
@@ -68,57 +69,67 @@ protected String getRpcCallbackUrl() {
...
+        }
+        throw new IOException(reason.getMessage(), reason); // can't be null here because logic in for loop
     }
 
-    /**
-     * Sends the message, retries if there was an error.
-     */
-    private synchronized Object[] sendMessage(int port, RpcRequest<String> request, int rpcRetryCounter)
-            throws IOException {
+    private byte[] send(int port, RpcRequest<String> request) throws Exception {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java:92]: Please do not throw `Exception`, either define your own like `HomematicException extends Exception` or use something like `IllegalStateException` (or what fits best) and catch only that one. The problem with catchin `Exception` is that it hides programming errors like NPE.

Comment (MHerbst) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java:92]: I agree with you regarding the catching and throwing of "Exception", but I tried to stay close to the original implementation and Jetty throws a lot of different exceptions. Will try to get rid of the Exception.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java [hunk 3]
```diff
@@ -78,57 +84,64 @@ protected String getRpcCallbackUrl() {
...
+                            respData.write(recvBuffer, 0, read);
                         }
-                        respData.write(recvBuffer, 0, read);
+                        ret = respData.toByteArray();
                     }
-                    ret = respData.toByteArray();
+                } else {
+                    logger.warn("Non-blocking XmlRpcRequest failed, status code: {} / request: {}", httpStatus,
+                            request);
+                    resp.abort(new Exception());
```
Comment (J-N-K) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java:136]:
```suggestion
                    resp.abort(new IOException());
```

Comment (MHerbst) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java]: Done

**Pull request merged**